### DISCUSSION TITLE
I want to work with multiple base url in retrofit2

### DISCUSSION THREAD

**Discussion by aumobiletools** [square/retrofit#3696] [Q&A]
I have a structure whose base url address is constantly variable. When my user changes his location in the application, I change the base url to show the closest server to him. but the retrofit client does not regenerate even though I restart the activities and change all the static variables. The only way to do this is to close the application completely and restart it.

I don't have a fixed base url. There could be 4 different locations today, 12 different locations tomorrow. so i need to find a permanent solution.

I have examined a lot in debug mode, but; When the base url changes, it never goes inside the getClient and cannot reassign the base url.

 

 
    GsonBuilder builder = new GsonBuilder();
    builder.registerTypeAdapter(Date.class, new JsonDeserializer<Date>() {
        @Override
        public Date deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
            return new Date(json.getAsJsonPrimitive().getAsLong());
        }
    });

    Gson gson = builder.create();
    OkHttpClient client = new OkHttpClient.Builder().addInterceptor(new Interceptor() {
        @Override
        public Response intercept(Chain chain) throws IOException {
            Request request = chain.request().newBuilder()
                    .addHeader("AUTHORIZATION", "" + JWT_TOKEN)
                    .build();
            return chain.proceed(request);
        }
    }).build();

    if (retrofit == null) {
        retrofit = new Retrofit.Builder().baseUrl(baseUrl)
                .addConverterFactory(GsonConverterFactory.create(gson))
                .client(client)
                .build();
        return retrofit;
    }

    return retrofit;

**Comment by craig-day** [comment]
I would implement a `CallFactory` that wraps your `client` instance. You can then override the url of the request object. 

```java
  @Override
  public Call newCall(@Nonnull Request request) {
    HttpUrl updatedUrl = request
        .url()
        .newBuilder()
        .host("my.new.host")
        .build();

    Request updatedRequest = request.newBuilder().url(updatedUrl).build();

    return delegateClient.newCall(updatedRequest);
  }
```

**Comment by gold24park** [comment]
I resolved it by create a new interceptor which can change host url. 
```
class HostInterceptor() : Interceptor {

    @Volatile
    private var host = ""

    override fun intercept(chain: Interceptor.Chain): Response {
        host = // your host...
        var request = chain.request()
        host.toHttpUrlOrNull()?.let {
            if (<Your condition>) {
                request = request.newBuilder()
                    .url(request.url.toUrl().toString().replace(request.url.host, it.host))
                    .build()
            }
        }
        return chain.proceed(request)
    }
}

```
